core/zoe/zoe/zoe.js

```diff
@@ -0,0 +1,331 @@
+       * invariants hold.
+       * @param  {object[]} offerIds - an array of offerIds
+       */
+      complete: async offerIds => {
+        const extents = readOnlyState.getExtentsFor(offerIds);
+        const extentOps = readOnlyState.getExtentOps(instanceId);
+        const labels = readOnlyState.getLabels(instanceId);
+        const assetDescs = toAssetDescMatrix(extentOps, labels, extents);
+        const purses = adminState.getPurses(instanceId);
+        const payments = await makePayments(purses, assetDescs);
+        const results = adminState.getResultsFor(offerIds);
+        results.map((result, i) => result.res(payments[i]));
+        adminState.removeOffers(offerIds);
+      },
+
+      /**
+       *  The governing contract can create an empty offer and get
+       *  the associated offerId. This allows the governing contract
+       *  to use this offer slot for recordkeeping. For instance, to
+       *  represent a pool, the governing contract can create an
+       *  empty offer and then reallocate other extents to this offer.
+       */
+      escrowEmptyOffer: length => {
+        // attenuate the authority by not passing along the result
+        // promise object and only passing the offerId
+        const { offerId } = escrowEmptyOffer(adminState.recordOffer, length);
+        return offerId;
+      },
+      /**
+       *  The governing contract can also create a real offer and
+       *  get the associated offerId, bypassing the seat and receipt
+       *  creation. This allows the governing contract to make
+       *  offers on the users' behalf, as happens in the
+       *  `addLiquidity` step of the `autoswap` contract.
+       */
+      escrowOffer: async (offerDesc, offerPayments) => {
+        // attenuate the authority by not passing along the result
+        // promise object and only passing the offerId
+        const { offerId } = await escrowOffer(
+          adminState.recordOffer,
+          adminState.getOrMakePurseForAssay,
+          offerDesc,
+          offerPayments,
+        );
+        return offerId;
+      },
+
+      burnEscrowReceipt: async escrowReceipt => {
+        const assetDesc = await escrowReceiptAssay.burnAll(escrowReceipt);
+        const { id } = assetDesc.extent;
+        const offerIds = harden([id]);
+        await fillInUndefinedExtents(
+          adminState,
+          readOnlyState,
+          offerIds,
+          instanceId,
+        );
+        adminState.recordUsedInInstance(instanceId, id);
+        return assetDesc.extent;
+      },
+
+      makeInvite: (offerToBeMade, useObj) => {
+        const inviteExtent = harden({
+          id: harden({}),
+          instanceId,
+          offerToBeMade,
+        });
+        const invitePurseP = inviteMint.mint(inviteExtent);
+        inviteAddUseObj(inviteExtent.id, useObj);
+        const invitePaymentP = invitePurseP.withdrawAll();
+        return invitePaymentP;
+      },
```

Comment:
This is the other side of the higher-order assayability issue (aka the Fred problem). The `inviteExtent` must contain all the information that Fred provides to the generic swapper (aka escrow exchange agent), to ensure that Alice cannot cheat him --- cannot sell him something other than the coveredCall he wants.

For the coveredCall, where would the (timing authority, deadline) go? Since the instanceId is specific to an instance, and Fred can credibly obtain instance info, perhaps we extend instances with terms?

Or perhaps the `offerToBeMade` parameter here is misnamed, and it is more like the contractHost's seatDesc?